teachertool/src/components/MakecodeFrame.tsx

```diff
@@ -1,140 +1,31 @@
 /// <reference path="../../../built/pxteditor.d.ts" />
 import * as React from "react";
-import { useState, useEffect, useRef } from "react";
-import { ProgressBar } from "react-common/components/controls/ProgressBar";
+import { useEffect, useRef } from "react";
+import { getIframeRef, removeIframeRef } from "../services/makecodeEditorService";
 interface MakeCodeFrameProps {
     pageSourceUrl: string;
-    highContrast?: boolean;
-    onFrameOpen: () => void;
-    onFrameClose: () => void;
     tutorialEventHandler?: (event: pxt.editor.EditorMessageTutorialEventRequest) => void;
 }
 
-type FrameState = "loading" | "no-project" | "opening-project" | "project-open" | "closing-project";
-
-interface PendingMessage {
-    original: pxt.editor.EditorMessageRequest;
-    handler: (response: any) => void;
-}
-
 export const MakeCodeFrame: React.FC<MakeCodeFrameProps> =
-    ( { pageSourceUrl,
-        highContrast,
-        onFrameClose,
-        onFrameOpen
-    } ) => {
-    let ref: HTMLIFrameElement | undefined;
-    const messageQueue: pxt.editor.EditorMessageRequest[] = [];
-    let nextId: number = 0;
-    let pendingMessages: {[index: string]: PendingMessage} = {};
-    const [frameState, setFrameState] = useState<FrameState>("loading");
-    const [loadPercent, setLoadPercent] = useState(0);
-    const [workspaceReady, setWorkspaceReady] = useState(false);
+    ( { pageSourceUrl} ) => {
 
     let frameRef = useRef<HTMLIFrameElement>(null);
```

Comment:
I strongly recommend using a method to receive the ref, like this:
```
const handleIFrameRef = (el: HTMLIFrameElement | null) => {
    EditorService.setEditorRef(el ?? undefined)
}
```

then:
```
<iframe ... ref={handleIFrameRef}></iframe>
```


This creates an airtight coupling in case the ref changes at any point during rerender.

Reply:
Does using React's useRef handle this?

Reply:
This way is a bit confusing to me. If we have the function be the ref for the iframe, the function wouldn't have anything as an argument so the argument for `setEditorRef` would always be undefined, wouldn't it?

Reply:
This is using some React functionality called a "callback ref". When the ref is set this way, React will automatically call the function you provide with the `HTMLIFrameElement` parameter set to the actual element in question. So in this case, `el` would point to the iframe node, and that's what you'd end up passing into `setEditorRef`.

https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs

Reply:
Oh, cool, thanks!

Reply:
> Does using React's useRef handle this?

The ref will receive the updated value, but it won't propagate anywhere. For propagation, the defined `useEffect` would need to be triggered when the ref changes. You might think this could be done by adding ref.current as a dependency, but this is problematic due to the ordering of updates in the component. ref values are updated after effect dependencies are evaluated. Because of this ordering, the React team considers refs in dependency lists an anti-pattern and it will be flagged by react/exhaustive-deps.

Solution is to use a callback ref (thanks for the doc link!). Here's a similar example from the React docs: https://legacy.reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node

And I see they're memoizing the callback ref with `useCallback`. Nice optimization. We could safely do that.